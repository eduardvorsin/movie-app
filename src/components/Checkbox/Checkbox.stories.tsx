import type { Meta, StoryObj } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import Checkbox, { Props } from './Checkbox';
import { ChangeEventHandler, useState } from 'react';

type Sizes = Exclude<Props['size'], undefined>;
type Variant = {
	id: string,
	isDisabled?: Props['isDisabled'],
	isRequired?: Props['isRequired'],
	isInvalid?: Props['isInvalid'],
	error?: Props['error'],
};

const sizes: Sizes[] = ['small', 'medium', 'large', 'xlarge'];
const variants: Variant[] = [
	{
		id: '0',
		isDisabled: true,
	},
	{
		id: '1',
		isInvalid: true,
	},
	{
		id: '2',
		isRequired: true,
	},
	{
		id: '3',
		error: 'error message',
	},
];

const meta: Meta<typeof Checkbox> = {
	title: 'components/Checkbox',
	tags: ['autodocs'],
	argTypes: {
		className: {
			description: 'Additional classes for the Checkbox'
		},
		testId: {
			description: 'Id for testing the component',
		},
		label: {
			description: 'Text description for the input field',
		},
		value: {
			description: 'Input value',
		},
		name: {
			description: 'Name for the input',
		},
		error: {
			description: 'Error Message',
		},
		id: {
			description: 'Input id',
		},
		isChecked: {
			description: 'Determines whether the checkbox is marked or not',
		},
		size: {
			description: 'Allows you to set the sizes of the checkbox',
		},
		isDisabled: {
			description: 'Disabling input makes it non-functional',
		},
		isInvalid: {
			description: 'Visually highlights the input in red when incorrectly entered data',
		},
		isRequired: {
			description: 'Makes input mandatory and adds a red asterisk at the end of the label',
		},
		onFocus: {
			description: 'The callback is executed when the focus hits the input',
			control: false,
		},
		onBlur: {
			description: 'The callback is executed when the input loses focus',
			control: false,
		},
		onChange: {
			description: 'The callback is executed when the input value changes',
			control: false,
		},
	},
	component: Checkbox,
	parameters: {
		docs: {
			description: {
				component: 'Checkboxes provide individuals with a means to choose one or multiple items from a set, or toggle between two options that are mutually exclusive (either selected or unselected).',
			},
		},
	},
}

export default meta;
type Story = StoryObj<typeof Checkbox>;

const CheckboxWithHooks = (props: Omit<Props, 'onChange' | 'isChecked'>) => {
	const [isChecked, setIsChecked] = useState<boolean>(false);

	const changeHandler: ChangeEventHandler<HTMLInputElement> = (e) => {
		setIsChecked(e.currentTarget.checked);
		action('Changed')(e);
	}

	return (
		<Checkbox
			{...props}
			isChecked={isChecked}
			onChange={changeHandler}
		/>
	)
};

export const Default: Story = {
	args: {
		id: 'default',
		name: 'default',
		label: 'default',
		value: 'default',
	},
	parameters: {
		docs: {
			description: {
				story: 'Utilize a CheckboxInput to enable users to make a variety of selections, including the option to choose zero, one, or multiple items.',
			},
		},
	},
	render: (args) => (<CheckboxWithHooks {...args} />),
};

export const Disabled: Story = {
	args: {
		id: 'disabled',
		name: 'disabled',
		label: 'disabled',
		value: 'disabled',
		isDisabled: true,
	},
	parameters: {
		docs: {
			description: {
				story: 'Implement this for actions that are presently inaccessible. The user interface should clearly communicate why the input is inactive and provide guidance on the necessary steps to activate it.',
			},
		},
	},
	render: (args) => (<CheckboxWithHooks {...args} />),
};

export const Required: Story = {
	args: {
		id: 'required',
		name: 'required',
		label: 'required',
		value: 'required',
		isRequired: true,
	},
	parameters: {
		docs: {
			description: {
				story: 'This is employed to signify that the input must be completed or filled out.',
			},
		},
	},
	render: (args) => (<CheckboxWithHooks {...args} />),
};

export const Invalid: Story = {
	args: {
		id: 'invalid',
		name: 'invalid',
		label: 'invalid',
		value: 'invalid',
		isInvalid: true,
	},
	parameters: {
		docs: {
			description: {
				story: 'It is utilized to visually emphasize the input by turning it red when it fails validation.',
			},
		},
	},
	render: (args) => (<CheckboxWithHooks {...args} />),
};

export const Error: Story = {
	args: {
		id: 'error',
		name: 'error',
		label: 'error',
		value: 'error',
		error: 'error message',
	},
	parameters: {
		docs: {
			description: {
				story: 'Shows an error message beneath the input field.',
			},
		},
	},
	render: (args) => (<CheckboxWithHooks {...args} />),
};

export const EventCallbacks: Story = {
	argTypes: {
		onFocus: {
			action: 'Focused',
		},
		onBlur: {
			action: 'Blured',
		},
	},
	args: {
		name: 'event-callbacks',
		id: 'event-callbacks',
		value: 'event-callbacks',
		label: 'focus, blur event',
	},
	parameters: {
		docs: {
			description: {
				story: 'The callback function can be provided to the onFocus, onBlur event.',
			},
		},
	},
	render: (args) => (<CheckboxWithHooks {...args} />)
};

export const Sizes: Story = {
	parameters: {
		docs: {
			description: {
				story: 'Shows an error message beneath the input field.',
			},
		},
	},
	render: () => (
		<div className='flex items-center'>
			{sizes.map((size) => (
				<CheckboxWithHooks
					value={size}
					key={size}
					className='mr-4 last:mr-4'
					id={size}
					name={size}
					label={size}
					size={size}
				/>
			))}
		</div>
	),
};

export const All: Story = {
	parameters: {
		docs: {
			description: {
				story: 'Here are all possible variants of avatars, in size, statuses, states, shape, etc.',
			},
		},
	},
	render: () => (
		<div
			className='grid gap-x-4 gap-y-6 grid-cols-[repeat(4,_minmax(0,_150px))]'
		>
			{sizes.map((size) => (
				variants.map((variant) => (
					<CheckboxWithHooks
						key={`${size}-${variant.id}`}
						id={`${size}-${variant.id}`}
						name={`${size}-${variant.id}`}
						label={`${size}-${variant.id}`}
						value={`${size}-${variant.id}`}
						size={size}
						isDisabled={variant.isDisabled}
						isInvalid={variant.isInvalid}
						isRequired={variant.isRequired}
						error={variant.error}
					/>
				))))}
		</div>
	),
};