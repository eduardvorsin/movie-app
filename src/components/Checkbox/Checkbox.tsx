'use client';
import { ChangeEventHandler, FocusEventHandler } from 'react';
import InlineMessage from '../InlineMessage/InlineMessage';

export type Props = {
	id: string,
	value: string,
	name: string,
	label: string,
	onChange: ChangeEventHandler<HTMLInputElement>,
	className?: string,
	testId?: string,
	size?: 'small' | 'medium' | 'large' | 'xlarge',
	error?: string,
	isChecked?: boolean,
	defaultChecked?: boolean,
	isDisabled?: boolean,
	isRequired?: boolean,
	isInvalid?: boolean,
	onFocus?: FocusEventHandler<HTMLInputElement>,
	onBlur?: FocusEventHandler<HTMLInputElement>,
};

const sizeTypes = {
	small: 'w-3 h-3',
	medium: 'w-4 h-4',
	large: 'w-5 h-5',
	xlarge: 'w-6 h-6',
}

export default function Checkbox({
	className,
	id,
	value,
	size = 'medium',
	name,
	label,
	testId,
	error,
	isDisabled,
	isChecked,
	isRequired,
	isInvalid,
	defaultChecked,
	onChange,
	onFocus,
	onBlur,
}: Props) {
	const labelClasses = [
		'inline-flex py-1 pl-5 items-center',
		isDisabled ? 'opacity-disabled cursor-not-allowed' : 'cursor-pointer',
	].join(' ');

	return (
		<div
			className={className}
		>
			<label
				className={labelClasses}
				data-testid={testId}
			>
				<input
					className='appearance-none absolute peer'
					type='checkbox'
					id={id}
					name={name}
					disabled={isDisabled}
					onChange={onChange}
					onFocus={onFocus}
					onBlur={onBlur}
					value={value}
					checked={isChecked}
					defaultChecked={defaultChecked}
					required={isRequired}
					aria-invalid={isInvalid}
					aria-describedby={error ? `${id}-error-message` : undefined}
					data-invalid={isInvalid}
				/>
				<svg
					className={`${sizeTypes[size]} -ml-5 rounded-1 border-2 border-neutral-500 fill-none peer-checked:bg-blue-700 peer-checked:fill-neutral-0 peer-checked:border-blue-500 peer-focus:border-blue-500 peer-data-[invalid]:border-red-600 transition-colors duration-100 ease-linear dark:peer-data-[invalid]:border-red-500 dark:peer-checked:bg-blue-400 dark:peer-focus:border-blue-300 dark:peer-checked:fill-dark-neutral-0 dark:peer-checked:border-blue-300 dark:border-dark-neutral-800`}
					viewBox='0 0 20 20'
				>
					<path
						fillRule='evenodd'
						d='M15.78 5.97a.75.75 0 0 1 0 1.06l-6.5 6.5a.75.75 0 0 1-1.06 0l-3.25-3.25a.75.75 0 1 1 1.06-1.06l2.72 2.72 5.97-5.97a.75.75 0 0 1 1.06 0Z'
					/>
				</svg>
				<span
					className='ml-2 text-100 text-dark-neutral-0 dark:text-neutral-0'
				>
					{label}
					{isRequired && (
						<span
							className='pl-0.5 text-red-800 dark:text-red-300'
							aria-hidden='true'
						>
							*
						</span>
					)}
				</span>
			</label>
			{error && (
				<InlineMessage
					className='mt-1'
					message={error}
					fieldId={`${id}-error-message`}
					appearance='error'
				/>
			)}
		</div>
	);
};

