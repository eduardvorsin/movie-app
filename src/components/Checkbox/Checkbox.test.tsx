import { render, screen } from "@testing-library/react";
import Checkbox from "./Checkbox";
import userEvent from "@testing-library/user-event";

describe('Checkbox tests', () => {
	it('is rendered correctly', () => {
		render(
			<Checkbox
				id='test-checkbox'
				name='test-checkbox'
				label='test-checkbox'
				value=''
				onChange={() => { }}
				isChecked={true}
			/>
		);

		expect(screen.getByRole<HTMLInputElement>('checkbox')).toBeInTheDocument();
	});

	it('when the focus is received, the mock function is triggered', async () => {
		const user = userEvent.setup();
		const mockFn = jest.fn();

		render(
			<Checkbox
				id='test-checkbox'
				name='test-checkbox'
				label='test-checkbox'
				value=''
				onChange={() => { }}
				onFocus={mockFn}
				isChecked={true}
			/>
		);

		await user.tab();

		expect(mockFn).toHaveBeenCalledTimes(1);
	});


	it('when the focus is lost, the mock function is triggered', async () => {
		const user = userEvent.setup();
		const mockFn = jest.fn();

		render(
			<Checkbox
				id='test-checkbox'
				name='test-checkbox'
				label='test-checkbox'
				value=''
				onChange={() => { }}
				onBlur={mockFn}
				isChecked={true}
			/>
		);

		await user.tab();
		await user.click(document.body);

		expect(mockFn).toHaveBeenCalledTimes(1);
	});

	it('when clicked, the mock function is triggered', async () => {
		const user = userEvent.setup();
		const mockFn = jest.fn();

		render(
			<Checkbox
				id='test-checkbox'
				name='test-checkbox'
				label='test-checkbox'
				value=''
				onChange={mockFn}
				isChecked={true}
			/>
		);

		await user.click(screen.getByRole<HTMLInputElement>('checkbox'));

		expect(mockFn).toHaveBeenCalledTimes(1);
	});

	it('if the value in the error sample is not empty, then the error element should be displayed.', () => {
		render(
			<Checkbox
				id='test-checkbox'
				name='test-checkbox'
				label='test-checkbox'
				value=''
				error='mock error'
				onChange={() => { }}
				isChecked={true}
			/>
		);

		expect(screen.getByText<HTMLParagraphElement>(/mock error/i)).toBeInTheDocument();
	});

	it('if isDisabled is true then the checkbox becomes inactive', async () => {
		const user = userEvent.setup();
		const mockFn = jest.fn();

		render(
			<Checkbox
				id='test-checkbox'
				name='test-checkbox'
				label='test-checkbox'
				value=''
				onChange={() => { }}
				isChecked={true}
			/>
		);

		await user.click(screen.getByRole<HTMLInputElement>('checkbox'));

		expect(mockFn).toHaveBeenCalledTimes(0);
	});

	it('if isRequired is true then an asterisk should appear in the label', () => {

		render(
			<Checkbox
				id='test-checkbox'
				name='test-checkbox'
				label='test-checkbox'
				value=''
				isRequired
				onChange={() => { }}
				isChecked={true}
			/>
		);

		expect(screen.getByText<HTMLSpanElement>('*')).toBeInTheDocument();
	});

	it('is a basic snapshot', () => {
		render(
			<Checkbox
				id='test-checkbox'
				name='test-checkbox'
				label='test-checkbox'
				value=''
				onChange={() => { }}
				isChecked={true}
			/>
		);

		expect(screen.getByRole<HTMLInputElement>('checkbox')).toMatchSnapshot();
	});

	it('is a snapshot with the "small" size type', () => {
		render(
			<Checkbox
				id='test-checkbox'
				name='test-checkbox'
				label='test-checkbox'
				value=''
				size='small'
				onChange={() => { }}
				isChecked={true}
			/>
		);

		expect(screen.getByRole<HTMLInputElement>('checkbox')).toMatchSnapshot();
	});

	it('is a snapshot with the "medium" size type', () => {
		render(
			<Checkbox
				id='test-checkbox'
				name='test-checkbox'
				label='test-checkbox'
				value=''
				size='medium'
				onChange={() => { }}
				isChecked={true}
			/>
		);

		expect(screen.getByRole<HTMLInputElement>('checkbox')).toMatchSnapshot();
	});

	it('is a snapshot with the "large" size type', () => {
		render(
			<Checkbox
				id='test-checkbox'
				name='test-checkbox'
				label='test-checkbox'
				value=''
				size='large'
				onChange={() => { }}
				isChecked={true}
			/>
		);

		expect(screen.getByRole<HTMLInputElement>('checkbox')).toMatchSnapshot();
	});

	it('is a snapshot with the "xlarge" size type', () => {
		render(
			<Checkbox
				id='test-checkbox'
				name='test-checkbox'
				label='test-checkbox'
				value=''
				size='xlarge'
				onChange={() => { }}
				isChecked={true}
			/>
		);

		expect(screen.getByRole<HTMLInputElement>('checkbox')).toMatchSnapshot();
	});

	it('is a snapshot with error element', () => {
		render(
			<Checkbox
				id='test-checkbox'
				name='test-checkbox'
				label='test-checkbox'
				value=''
				error='mock error'
				onChange={() => { }}
				isChecked={true}
			/>
		);

		expect(screen.getByRole<HTMLInputElement>('checkbox')).toMatchSnapshot();
	});

	it('is snapshot with isRequired equal to true', () => {
		render(
			<Checkbox
				id='test-checkbox'
				name='test-checkbox'
				label='test-checkbox'
				value=''
				isRequired
				onChange={() => { }}
				isChecked={true}
			/>
		);

		expect(screen.getByRole<HTMLInputElement>('checkbox')).toMatchSnapshot();
	});

	it('is snapshot with isDisabled equal to true', () => {
		render(
			<Checkbox
				id='test-checkbox'
				name='test-checkbox'
				label='test-checkbox'
				value=''
				isDisabled
				onChange={() => { }}
				isChecked={true}
			/>
		);

		expect(screen.getByRole<HTMLInputElement>('checkbox')).toMatchSnapshot();
	});

	it('is snapshot with isInvalid equal to true', () => {
		render(
			<Checkbox
				id='test-checkbox'
				name='test-checkbox'
				label='test-checkbox'
				isInvalid
				value=''
				onChange={() => { }}
				isChecked={true}
			/>
		);

		expect(screen.getByRole<HTMLInputElement>('checkbox')).toMatchSnapshot();
	});
});