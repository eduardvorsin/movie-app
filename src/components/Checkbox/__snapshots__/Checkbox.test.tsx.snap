// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox tests is a basic snapshot 1`] = `
<input
  checked=""
  class="appearance-none absolute peer"
  id="test-checkbox"
  name="test-checkbox"
  type="checkbox"
  value=""
/>
`;

exports[`Checkbox tests is a snapshot with error element 1`] = `
<input
  aria-describedby="test-checkbox-error-message"
  checked=""
  class="appearance-none absolute peer"
  id="test-checkbox"
  name="test-checkbox"
  type="checkbox"
  value=""
/>
`;

exports[`Checkbox tests is a snapshot with the "large" size type 1`] = `
<input
  checked=""
  class="appearance-none absolute peer"
  id="test-checkbox"
  name="test-checkbox"
  type="checkbox"
  value=""
/>
`;

exports[`Checkbox tests is a snapshot with the "medium" size type 1`] = `
<input
  checked=""
  class="appearance-none absolute peer"
  id="test-checkbox"
  name="test-checkbox"
  type="checkbox"
  value=""
/>
`;

exports[`Checkbox tests is a snapshot with the "small" size type 1`] = `
<input
  checked=""
  class="appearance-none absolute peer"
  id="test-checkbox"
  name="test-checkbox"
  type="checkbox"
  value=""
/>
`;

exports[`Checkbox tests is a snapshot with the "xlarge" size type 1`] = `
<input
  checked=""
  class="appearance-none absolute peer"
  id="test-checkbox"
  name="test-checkbox"
  type="checkbox"
  value=""
/>
`;

exports[`Checkbox tests is snapshot with isDisabled equal to true 1`] = `
<input
  checked=""
  class="appearance-none absolute peer"
  disabled=""
  id="test-checkbox"
  name="test-checkbox"
  type="checkbox"
  value=""
/>
`;

exports[`Checkbox tests is snapshot with isInvalid equal to true 1`] = `
<input
  aria-invalid="true"
  checked=""
  class="appearance-none absolute peer"
  data-invalid="true"
  id="test-checkbox"
  name="test-checkbox"
  type="checkbox"
  value=""
/>
`;

exports[`Checkbox tests is snapshot with isRequired equal to true 1`] = `
<input
  checked=""
  class="appearance-none absolute peer"
  id="test-checkbox"
  name="test-checkbox"
  required=""
  type="checkbox"
  value=""
/>
`;
